runtime/src/trace.rs

```diff
@@ -0,0 +1,498 @@
+                    .get_mut::<OtelData>()
+                    .and_then(|data| {
+                        data.builder
+                            .trace_id
+                            .or(Some(TraceId::INVALID))
+                            .zip(data.builder.span_id.or(Some(SpanId::INVALID)))
+                    })
+            })
+        {
+            record.set_trace_context(trace_id, span_id, None);
+        }
+
+        attributes.0.extend(
+            ctx.event_scope(event)
+                .into_iter()
+                .flatten()
+                .flat_map(
+                    |span| match span.extensions_mut().get_mut::<SpanFieldValues>() {
+                        Some(values) if !values.0.is_empty() => Some(values.clone().0.into_iter()),
+                        _ => None,
+                    },
+                )
+                .flatten()
+                .chain(
+                    [
+                        metadata.file().map(|file| {
+                            (
+                                opentelemetry::Key::from(CODE_FILEPATH),
+                                opentelemetry::logs::AnyValue::from(file.to_string()),
+                            )
+                        }),
+                        metadata.line().map(|line| {
+                            (
+                                opentelemetry::Key::from(CODE_LINENO),
+                                opentelemetry::logs::AnyValue::from(line),
+                            )
+                        }),
+                        metadata.module_path().map(|path| {
+                            (
+                                opentelemetry::Key::from("code.module_path"),
+                                opentelemetry::logs::AnyValue::from(path.to_string()),
+                            )
+                        }),
+                    ]
+                    .into_iter()
+                    .flatten(),
+                ),
+        );
+
+        record.add_attributes(attributes.0);
+
+        self.logger.emit(record)
+    }
+}
+
+/// Convert a [`log::Level`] to its equivalent [`Severity`]
+#[inline(always)]
+pub(crate) fn log_level_as_severity(level: log::Level) -> Severity {
+    match level {
+        log::Level::Info => Severity::Info,
+        log::Level::Warn => Severity::Warn,
+        log::Level::Debug => Severity::Debug,
+        log::Level::Trace => Severity::Trace,
+        log::Level::Error => Severity::Error,
+    }
+}
+
+// Create a Resource that captures information about the entity for which telemetry is recorded.
+pub fn resource(crate_name: &'static str, package_version: &'static str) -> Resource {
+    Resource::from_schema_url(
+        [
+            Some(KeyValue::new(SERVICE_NAME, crate_name)),
+            Some(KeyValue::new(SERVICE_VERSION, package_version)),
+            Some(KeyValue::new(TELEMETRY_SDK_NAME, "opentelemetry")),
+            Some(KeyValue::new(TELEMETRY_SDK_VERSION, "0.27.1")),
```

Comment:
Should we set this from a constant, with some notes to remind us to update it when we update the sdk?

Reply:
[Working](https://github.com/open-telemetry/opentelemetry-rust/pull/2504) on it 🥲

Reply:
A comment in Cargo.toml should do fine.

Reply:
Otel guy mentioned it is populated automatically. Did you observe otherwise?
Anyways, don't let this block. We can have it hardcoded until your otel PR is resolved.

Reply:
Yeah, it's not populated *at all* unless you do it here manually. Unless he means it's populated... somewhere else 🤔? Anything we could *not* do here would be nice. Do you have a link/reference to where he mentioned it?

Reply:
> A comment in Cargo.toml should do fine.

[Added](https://github.com/shuttle-hq/shuttle/blob/feat/user-project-telemetry/runtime/Cargo.toml#L31)